src/cudadev/CUDADataFormats/SiPixelDigisCUDA.h

```diff
@@ -26,60 +78,49 @@ class SiPixelDigisCUDA {
   uint32_t nModules() const { return nModules_h; }
   uint32_t nDigis() const { return nDigis_h; }
 
-  uint16_t *xx() { return xx_d.get(); }
-  uint16_t *yy() { return yy_d.get(); }
-  uint16_t *adc() { return adc_d.get(); }
-  uint16_t *moduleInd() { return moduleInd_d.get(); }
-  int32_t *clus() { return clus_d.get(); }
-  uint32_t *pdigi() { return pdigi_d.get(); }
-  uint32_t *rawIdArr() { return rawIdArr_d.get(); }
-
-  uint16_t const *xx() const { return xx_d.get(); }
-  uint16_t const *yy() const { return yy_d.get(); }
-  uint16_t const *adc() const { return adc_d.get(); }
-  uint16_t const *moduleInd() const { return moduleInd_d.get(); }
-  int32_t const *clus() const { return clus_d.get(); }
-  uint32_t const *pdigi() const { return pdigi_d.get(); }
-  uint32_t const *rawIdArr() const { return rawIdArr_d.get(); }
-
-  cms::cuda::host::unique_ptr<uint16_t[]> adcToHostAsync(cudaStream_t stream) const;
-  cms::cuda::host::unique_ptr<int32_t[]> clusToHostAsync(cudaStream_t stream) const;
-  cms::cuda::host::unique_ptr<uint32_t[]> pdigiToHostAsync(cudaStream_t stream) const;
-  cms::cuda::host::unique_ptr<uint32_t[]> rawIdArrToHostAsync(cudaStream_t stream) const;
-
-  class DeviceConstView {
+  uint16_t *xx() { return deviceFullView_.xx(); }
+  uint16_t *yy() { return deviceFullView_.yy(); }
+  uint16_t *adc() { return deviceFullView_.adc(); }
+  uint16_t *moduleInd() { return deviceFullView_.moduleInd(); }
+  int32_t *clus() { return deviceFullView_.clus(); }
+  uint32_t *pdigi() { return deviceFullView_.pdigi(); }
+  uint32_t *rawIdArr() { return deviceFullView_.rawIdArr(); }
+
+  uint16_t const *xx() const { return deviceFullView_.xx(); }
+  uint16_t const *yy() const { return deviceFullView_.yy(); }
+  uint16_t const *adc() const { return deviceFullView_.adc(); }
+  uint16_t const *moduleInd() const { return deviceFullView_.moduleInd(); }
+  int32_t const *clus() const { return deviceFullView_.clus(); }
+  uint32_t const *pdigi() const { return deviceFullView_.pdigi(); }
+  uint32_t const *rawIdArr() const { return deviceFullView_.rawIdArr(); }
+
+  class HostStoreAndBuffer {
+    friend SiPixelDigisCUDA;
   public:
-    __device__ __forceinline__ uint16_t xx(int i) const { return __ldg(xx_ + i); }
-    __device__ __forceinline__ uint16_t yy(int i) const { return __ldg(yy_ + i); }
-    __device__ __forceinline__ uint16_t adc(int i) const { return __ldg(adc_ + i); }
-    __device__ __forceinline__ uint16_t moduleInd(int i) const { return __ldg(moduleInd_ + i); }
-    __device__ __forceinline__ int32_t clus(int i) const { return __ldg(clus_ + i); }
-
-    uint16_t const *xx_;
-    uint16_t const *yy_;
-    uint16_t const *adc_;
-    uint16_t const *moduleInd_;
-    int32_t const *clus_;
+    HostStoreAndBuffer();
+    const SiPixelDigisCUDA::HostDeviceStore store() { return hostStore_; }
```

Comment:
I'd suggest to remove the `const` because it doesn't really enforce anything here
```suggestion
    SiPixelDigisCUDA::HostDeviceStore store() { return hostStore_; }
```

Reply:
Unless I miss something, it does. Data access functions have const/non-const variants. I made a macro expansion of this class here: https://gist.github.com/ericcano/1d10e0902829408a0d8815e99567102c

Reply:
I can think of it preventing e.g. the following
```cpp
SiPixelDigisCUDA const& digis = ctx.get(iEvent, ...);
auto hostData = digis.dataToHostAsync(ctx.stream());
hostData.store().adc(0) = 42;
```
but it allows
```cpp
SiPixelDigisCUDA const& digis = ctx.get(iEvent, ...);
auto hostData = digis.dataToHostAsync(ctx.stream());
SiPixelDigisCUDA::HostDeviceStore store = hostData.store();
store.adc(0) = 42;
```

On one hand in this case the function itself is not `const`, so returning non-`const` would be fine.

Reply:
Thanks, I tried to turn it around, but indeed, we have no way to prevent this while retaining the trivial copyability, which is not only nice but also probably needed to pass view/store as arguments to a kernel (they get copied then).
So we do have to create a stripped down, const-only flavor of the store. Having a one-way copy constructor (non-const -> const) might simplify the coding and should not be too much hassle as it's few lines in the macros (and this would apply also to view).

Reply:
In the end, the store (renamed to layout in the follow up PR) is exclusively taking care of the columns layout. The access is then done via views (with const and mutable flavors).